fix NoProof block signature

winsvega opened the pull request:
Other devs report that It is better if NoProof blocks still have the signature. but the signature is just ignored. 
Rather then absence of signature as it is now.

chfast commented on test/tools/libtesteth/BlockChainHelper.cpp:
```diff
@@ -273,6 +273,14 @@ void TestBlock::mine(TestBlockChain const& _bc)
         recalcBlockHeaderBytes();
+
+    // Create dummy signature for NoProof blocks
+    if (Ethash::mixHash(m_blockHeader) == h256(0))
```
`h256{0}`.

chfast commented on test/tools/libtesteth/BlockChainHelper.cpp:
```diff
@@ -273,6 +273,14 @@ void TestBlock::mine(TestBlockChain const& _bc)
+    // Create dummy signature for NoProof blocks
+    if (Ethash::mixHash(m_blockHeader) == h256(0))
+    {
+        m_blockHeader.setSeal(Ethash::NonceField, h64(42));
```
Maybe make it null?

gumb0 commented on test/tools/libtesteth/BlockChainHelper.cpp:
```diff
@@ -273,6 +273,14 @@ void TestBlock::mine(TestBlockChain const& _bc)
+    if (Ethash::mixHash(m_blockHeader) == h256{0})
+    {
+        m_blockHeader.setSeal(Ethash::NonceField, h64{0});
+        m_blockHeader.setSeal(Ethash::MixHashField, h256{0});
```
Why do you assign it here if you just checked that it's 0?

winsvega replied:
because it returns 0 also when there is no signature set.

winsvega replied:
We could implenent a method hasSignature() if you woukd like. That will change libethereum. By default I try not to touch libethereum for test related logic

gumb0 commented on test/tools/jsontests/BlockChainTests.cpp:
```diff
@@ -109,7 +109,7 @@ json_spirit::mValue BlockchainTestSuite::doTests(json_spirit::mValue const& _inp
                     json_spirit::mObject const& expectObj = expect.get_obj();
                     ImportTest::parseJsonStrValueIntoSet(expectObj.at("network"), netlist);
-
+                    netlist = test::translateNetworks(netlist);
```
Did you just sneak in this unrelated change? :)

gumb0 commented:
I don't get - if there's a change to `NoProof` blocks format - why only testeth files are changed?
Will `test_mineBlocks` generate correct blocks with `NoProof` config?

gumb0 commented:
Do you mean both blocks witthout any signature and blocks with zero signature are "correct"?
This doesn't make sense to me, only one format should be valid (the one that other clients would also support)

winsvega commented:
Correct in what context? 
blocks without signature are incorrect for the test format. 
blocks with or without signature could still be invalid or correct depending on a block contents. 
blocks without signature or with invalid signature are incorret on Ethash

winsvega commented:
There is no consensus on blocks without signature cause NoProof is smth that we use for testing purposes. So we define what a block on NoProof should look like (with zero signature / random signature  or no signature at all) 
It is also possible to support blocks without signature in test format. It just that other clients devs agreed that it is easier for them to parse blocks with at least some signature.

gumb0 commented:
> So we define what a block on NoProof should look like (with zero signature / random signature or no signature at all)

So I understood that intention of this PR is to establish that NoProof blocks have zero signature.

> It just that other clients devs agreed that it is easier for them to parse blocks with at least some signature.

Right, so I'm saying we should generate blocks with signature everywhere.

> blocks with or without signature could still be invalid or correct depending on a block contents.

Why should blocks without signature be valid?

winsvega commented:
Blocks without signature are valid on NoProof

gumb0 commented:
But you say they are not valid for other clients

winsvega commented:
Other clients parsing state test format expect signature in blocks. Only in test format. They dont have NoProof defenition. We could make it anything.

gumb0 commented:
We could, but why would we want two different block formats in tests and out of the tests? This will only lead to confusion and possible incompatibilities between parts of the codebase.

chfast commented:
> Thats why this pr adding a zero signature to the blocks without signature

But you don't change NoProof design, just hack the final test output.

I'm also a bit into the solution to make NoProof more strict and do not allow block headers without signature fields.

winsvega commented:
By default I try not to touch core libraries logic for testing purposes. 
So If you think that NoProof should be modified on a core level to require signature at the block but not verifying it we could open another PR. 

To summarize the discussion: 
Is it concluded that NoProof should generate 0 signature at corelib mining method? 
Is it concluded that aleth should require signature to be present in a block rlp for any Ethash/NoProof

chfast commented:
Yes, if that makes the whole system more consistent and simple to understand.

Can someone check if Clique is adding/remove any fields to the block hash?

gumb0 commented:
@chfast It does not change the format, just changes the meaning of some fields https://github.com/ethereum/EIPs/issues/225

winsvega commented:
updatet this PR. now 0 signature is set by default inside sealEngineBase generateSeal method. 
that is used when NoProof is enabled.

gumb0 commented:
`bcBasicInsert` test is failing

gumb0 commented on libethcore/SealEngine.h:
```diff
@@ -99,8 +99,11 @@ class SealEngineBase: public SealEngineFace
 public:
 	void generateSeal(BlockHeader const& _bi) override
 	{
+        BlockHeader header(_bi);
```
I think this should happen in `NoProof`, not in `SealEngineBase`
Also I don't see any reason for `SealEngineBase` to have `generateSeal` implementation at all, so try to just move it to `NoProof`

winsvega commented:
removed generateSeal from SealEngineBase
moved generateSeal from SealEngineBase to NoProof.
Fixed bcBasicImport by moving the old implementation of generateSeal from SealEngineBase to BasicAuthority generateSeal

There should be an error throwing in case no callback is called from BasicAuthority::generateSeal. 
That leads to endless loop.

gumb0 commented:
Where is the endless loop?

gumb0 commented on libethcore/SealEngine.cpp:
```diff
@@ -29,6 +29,17 @@ void NoProof::init()
+void NoProof::generateSeal(BlockHeader const& _bi)
+{
+    BlockHeader header(_bi);
+    header.setSeal(1, h64{0});   // NonceField
```
Maybe move `enum { MixHashField = 0, NonceField = 1 };` to `SealEngineBase` and use it here

winsvega replied:
this is a good question. what if with sealEngine = Ethash2  there will be 3 or 1 fields for signature?

gumb0 replied:
We'll see then, let's not try to guess the future

winsvega commented:
Endless loop somewhere where it waits for a callback of m_onSealGenerated

gumb0 commented on libethcore/BasicAuthority.cpp:
```diff
@@ -50,8 +50,12 @@ void BasicAuthority::generateSeal(BlockHeader const& _bi)
+    bi.streamRLP(ret);
+    if (m_onSealGenerated)
+        m_onSealGenerated(ret.out());
+    // Else throw an error!!!
```
I'd say it's a problem of the calling code if it would not set this handler and then wait somehow (how?) for it to be called.
So I don't think anything need to be done here.
